src/basic/log.c

```diff
@@ -1352,6 +1360,14 @@ int log_syntax_internal(
                 const char *func,
                 const char *format, ...) {
 
+        int r;
+
+        if (log_syntax_callback) {
+                r = log_syntax_callback(unit, level, log_syntax_callback_userdata);
```

Comment:
We're calling set_put_strdup from within the callback function which could result in an out of memory error and in that case it would fail. Should we ignore it in that case?

Reply:
no, catch it, and propagate it back to the verify code. But the thing is, logging is assumed to be a side-effect, besides slowing things down it shouldn't affect program behaviour if we do everything right. Because we never check return errors of the log functions in general, since if we did the logging would totall change control flow if it wants, and it shouldn't do that.

This is a real problem btw. i.e. people in the past started a daemon of ours with some pipe as stderr, and then closed the pipe, and were surprised we'd die because there was some bug were we didnt properly ignore errors from writing to stderr. 

So yes, return values from logging should generally not be tested (i mean, how would ou rcode even look like if every fricking log_debug() code would mean we'd check its return value, the code size would explode).

But yes, you should handle the set_put_strdup() failure, but not by returning an error from your callback. Instead,  store this fact in userdata somehow. You could for example set free the set and then set it to POINTER_MAX (i.e. a special value that shall indicate: yikes, i had no memory). Then, after everything is done, you check if the set is POINTER_MAX before using it, and if so, you know something happened.

Reply:
> Would something like this work:
> 
> src/analyze/analyze-verify.c :
> 
> ```
>     static void log_syntax_callback(const char *unit, int level, void *userdata) {
>             Set **s = userdata;
>             int r;
> 
>             assert(userdata);
>             assert(unit);
>     
>             if (level <= LOG_WARNING) {
>                     r = set_put_strdup(s, unit);
>                     if (r < 0) {
>                             set_freep(s);
>                             userdata = POINTER_MAX;
>                     }
>             }
>     }
> ```

`set_free` rather than `set_freep`, and `s = POINTER_MAX` - userdata is local to the function, so it would not propagate back

> And then using that within src/basic/log.c :
> 
> ```
>     if (log_syntax_callback) {
>             log_syntax_callback(unit, level, log_syntax_callback_userdata);
>             if (log_syntax_callback_userdata == POINTER_MAX)
>                     return -ENOMEM;
>     }
> ```

The check should happen in verify_units, before the `set_isempty` call - the idea was to avoid aborting the log_syntax call, but instead have the owner of the callback do the verification

Reply:
set_free takes in type Set * whereas s is of type Set **